### ISSUE TITLE
Research and fix on GHA and upward procedure

### ISSUE THREAD

**Issue by MarcelKemp** [wazuh/wazuh#21933]
## Description

This issue aims to investigate and implement actions to minimize failed checks when updating main versioning branches, to automate possible actions to avoid failures and to fix inconsistent versioning cases in workflows. Examples of issues:
- https://github.com/wazuh/wazuh-packages/issues/2813
- https://github.com/wazuh/wazuh/issues/21797
- https://github.com/wazuh/wazuh-packages/issues/2823
- https://github.com/wazuh/wazuh/issues/21890

To this end, after a call with @jnasselle, we have compiled several points that we can look into to see the best available solution, which does not require too much time, because there are redesign issues where several of these problems should be addressed. These are:
- Create an order of steps explained in the templates of these issues, to minimize check failures due to missing branches, or the need to merge one repository before another because the GHA points to the target branch.
  - Ex: https://github.com/wazuh/wazuh-packages/issues/2763#issuecomment-1887478986
  - In addition, add new steps necessary to avoid future failed checks, as can be seen with the creation of new versions, as the dashboard and indexer packages for these new versions are not generated:
    - https://github.com/wazuh/wazuh-packages/pull/2832 - https://github.com/wazuh/wazuh-packages/runs/21718517214
- Modify the triggers of the checks, so that in the case of a new version being created, the corresponding pipelines are launched to generate the necessary content to cover the checks of that version. Also, it would be good to check if it is possible to launch pipelines just after detecting a merge, because in this way we could automatically launch some actions that need the branch to be merged.
  - https://github.com/wazuh/wazuh-packages/actions/workflows/upload-deb-images.yml 
  - https://github.com/wazuh/wazuh-packages/actions/workflows/upload-rpm-images.yml
  - https://github.com/wazuh/wazuh-packages/pull/2825
- Change the versioning format of the image creation and the building and use of these images, so that they match (using the specific version). As we have seen cases where there is a version discrepancy that causes it to fail (5.0 != latest).
  - [Upload package creation Docker images - DEB - amd64 and i386](https://github.com/wazuh/wazuh-packages/actions/runs/7886301675/job/21521766996) - Generate `deb_manager_builder_amd64:5.0` image.
  - [Build Wazuh Packages - DEB - amd64 and i386](https://github.com/wazuh/wazuh-packages/actions/runs/7887170791/job/21521814217) - Fail because try to match with `deb_manager_builder_amd64:latest`.

### Tasks

- [ ] Review current issues templates and update them, adding more details, steps and new instructions if necessary. 
- [x] Review check triggers and propose necessary changes for some issues.
- [x] Fix different version of images created and used.

**Comment by jnasselle** [comment]
# Update - WIP

## Current problems (with wazuh/wazuh and wazuh/wazuh-packages as different repos)

### Branch misalignment

#### wazuh/wazuh branch pickup 
- Description: currently wazuh/wazuh PRs checks (executed by Jenkins) attempt a "best match" branch selection using the next criteria:
- Jenkins reference: https://github.com/wazuh/wazuh-jenkins/blob/a17ff10d50205604f1d10b4b1cc807a96ff1abff/scripts/lambda/launchJenkinsJob.py#L138-L149
- Packages reference: https://github.com/wazuh/wazuh-jenkins/blob/a17ff10d50205604f1d10b4b1cc807a96ff1abff/jenkins-files/pull_requests/pr_wazuh.groovy#L142-L150
- QA reference: https://github.com/wazuh/wazuh-jenkins/blob/a17ff10d50205604f1d10b4b1cc807a96ff1abff/jenkins-files/pull_requests/pr_wazuh.groovy#L152-L160

TL;DR:
- Attempt to use a <target repo> branch name with the same PR branch
- Attempt to use the  <target repo> master branch if the current PR version matches it
- Attempt to use a  <target repo> branch name with the same long version of the current PR
- Attempt to use a  <target repo> branch name with the same short version of the current PR

#### wazuh/wazuh-packages branch pickup 
- Description: currently wazuh/wazuh-packages PRs checks (executed by GHA) attempt a the next logic to validate its changes against wazuh/wazuh code

- Wazuh reference:  https://github.com/wazuh/wazuh-packages/blob/8bf1bbbd9be9c622d06ddddc837f4cecea6dba6e/.github/workflows/build-deb-packages.yml#L53-L60

TL;DR:

- Container pick
  - Use container TAG being the container tag the commit-id of the PR if the PR touches container-related resources (and means that other flows of the same PR already built and uploaded the image)
  - Use container TAG being the PR's MAJOR.MINOR value if the PR did not touch container-related resources.
- wazuh/wazuh source code pick
  - Attempt to use the master branch if the current PR MAJOR.MINOR match 5.0 (FIXED VALUE)
  - Attempt to use the current PR MAJOR.MINOR.PATCH branch

**Comment by MarcelKemp** [comment]
### Update
- I am working on the procedure of steps needed to minimize failed checks, so that I can avoid failing some checks due to not having images created, updated or having deleted a branch.
  - Also taking into account the following: https://github.com/wazuh/wazuh-packages/pull/2868#issuecomment-1983672840
- On the other hand, I have detected that the problem of versioning and the use of `latest` happens in cases where the GHA is manually triggered. So, I think it is not necessary to modify this, as we are working on migrating and redesigning packages:
  - https://github.com/wazuh/wazuh/issues/21755
- To automate the generation of Docker images for package creation in GHA, it is necessary to modify the files that need to detect a change in order to trigger it. In addition, it is possible to run a workflow after merge a PR:
  - https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#running-your-pull_request-workflow-when-a-pull-request-merges

**Comment by MarcelKemp** [comment]
## Proposals

Firstly, I think it would be necessary to modify the templates for creating issues with the `github-branch-maintenance` objective, in order to organize the steps to follow and try to minimize failed checks to try to reduce the time.

Therefore, the templates that I consider most important or that I think need to be changed, would be:

### Wazuh-packages templates
- `Post release tasks for <version>` and `Scheduled upward merges for numbered branches in wazuh-packages repository`: 
  - From the merge to `4.9.0`, remove the lines for: `Update the Wazuh indexer and Wazuh dashboard ECR resources if needed`.
  - Reason: https://github.com/wazuh/wazuh-packages/pull/2868#issuecomment-1983672840
> - `child_branch_maintenance_post_release.txt`
> - `child_branch_maintenance_scheduled.txt`

### Wazuh templates
- For new versions coming out of the `master` branch, it is necessary to add a line before uploading the GHA images, in which we specify to update the version value, so that the build and generation of the GHA images will be executed correctly:
  - https://github.com/wazuh/wazuh-packages/pull/2825
  - In addition, we can configure it so that the `Build and push testing images for GitHub actions` step is done automatically in the previous PR, by modifying the triggers. 
> -  `child_support_version_minor.txt`

### Both templates
- Have the `wazuh` and `wazuh-packages` branches created in parallel, so that when generating the PR, the checks correctly take the branch and do not fail.


### Others
On the other hand, to avoid failures in the checks of wazuh-packages, we could modify the workflows of building and testing packages, to use the version or branch that best suits the situation, as discussed:
- https://github.com/wazuh/wazuh/issues/21933#issuecomment-1948361399
- The same will be taken into account with package redesign.
